### PR TITLE
Check for global.process before accessing

### PR DESCRIPTION
**Pull request by bre7**

### CODE REVIEW

**Comment by webmaster128** [comment]
Could you look into this? This is needed for using thick package in a webpack 5 project.

**Comment by webmaster128** [comment]
See 
Screenshot_2020-12-21 To v5 from v4 webpack from https://webpack.js.org/migrate/5/

File: lib/default-encoding.js [hunk 1]
```diff
@@ -1,8 +1,8 @@
 var defaultEncoding
 /* istanbul ignore next */
-if (process.browser) {
+if (global.process && global.process.browser) {
```
Comment (webmaster128) [inline lib/default-encoding.js:3]:
```suggestion
if (typeof process !== "undefined" && process.browser) {
```

File: lib/default-encoding.js [hunk 2]
```diff
@@ -1,8 +1,8 @@
...
   defaultEncoding = 'utf-8'
-} else if (process.version) {
+} else if (global.process && global.process.version) {
```
Comment (webmaster128) [inline lib/default-encoding.js:5]:
```suggestion
} else if (typeof process !== "undefined" && process.version) {
```

**Review by webmaster128 — COMMENTED** [review]
As far as I know, this is the common way to test the existence of globals without accessing them. However, the original version might work equally well.

**Comment by calvinmetcalf** [comment]
webpack decided to unilaterally break a ton of modules in v5 by intentionally removing the support it previously had for making a module work in both node and the browser. so for now consider this package to be incompatible with webpack 5 without configuring to put process back in.

**Comment by webmaster128** [comment]
@calvinmetcalf I see your concern here. I'm not familiar with the decision making process and to what degree it is justified. However, users of this library would be happy to be able to use it in a variety of environments, including Webpack 4, Webpack 5 or other bundlers. I wonder if we could refactor the library in a way that does not depend on a `process` global at all. `default-encoding.js` can return `utf-8` in all cases, assuming no support for node < 6 is required anymore. `process.nextTick` can be implemented via `setTimeout` and `browserPbkdf2` can be used when subtle is available. Would you be willing to accept a PR that removes all usages of `process`?

**Comment by webmaster128** [comment]
@calvinmetcalf there seems to be quite some support for the proposed path to webpack5 support and I would really like to avoid an unnecessary fork of this library just because people want to use it with webpack5. I would love to hear what you think about this approach.

File: lib/default-encoding.js (hunk 1 above)
Comment (calvinmetcalf) [reply lib/default-encoding.js:3]: does global not work in webpack 5 ?

Comment (webmaster128) [reply lib/default-encoding.js:3]: Good question. I wish I had written more to this suggestion when I wrote it last year.

But if you want to optionally access `process.browser` the proposed code is safe and does not rely on a non-standard `global` variable. [globalThis](https://developer.mozilla.org/de/docs/Web/JavaScript/Reference/Global_Objects/globalThis) will solve this in a standardized way for modern environments.

Comment (calvinmetcalf) [reply lib/default-encoding.js:3]: actually we have a BUNCH of places in here that reference global so if that's something that needs to be removed we should do that in a different pull

Comment (webmaster128) [reply lib/default-encoding.js:3]: Sure. But this PR introduces new usages of `global` that can be avoided.

**Comment by calvinmetcalf** [comment]
I'm not sure how helpful this will end up being as we also have some process.nextTick usage in lib/async.js, thoughts on fixing that @webmaster128 ?

**Comment by webmaster128** [comment]
> I'm not sure how helpful this will end up being as we also have some process.nextTick usage in lib/async.js, thoughts on fixing that @webmaster128 ?

`process.nextTick(callback)` can be replaced with `setTimeout(callback, 0)`, which is almost the same except for minimal implementation details that we most likely do not care about here.

**Comment by calvinmetcalf** [comment]
ugh, now I gotta write a little shim because webpack decided to stop actually supporting code that litterally runs both in the browser and node.

**Comment by webmaster128** [comment]
What do you need the shim for? All the node specific code I found could be rewritten to plain JavaScript without much trouble (https://github.com/crypto-browserify/pbkdf2/pull/92#issuecomment-755692497)

**Comment by webmaster128** [comment]
I am happy to create a PR for this if you would be willing to review and integrate.

**Comment by calvinmetcalf** [comment]
ok so one of the points behind this was to have it work with older versions of node, so I'd prefer not to drop any compatibility and if we do we'd need to do a version bump.

setTimeout does have some differences compared to process.nextTick (even when used in the browser) mainly when used recursively.  Like I could write some code with this library that would be much slower using setTimeout so I wrote a quick patch that picks the fastest thing available which I was about to push up about an hour ago before I got pulled into a meeting which I just added.

assuming this works in webpack 5 then that should fix the immediate problem at which point I'm totally open to refactoring this library in ways to make it more useful.

**Comment by calvinmetcalf** [comment]
ok I can publish this tommorow

**Comment by webmaster128** [comment]
Thats a lot. Here you find a demo showing this package now works with webpack 5: https://github.com/cosmos/cosmjs/issues/578#issuecomment-818615647

**Pull request merged**